### PR TITLE
GenAI Layer Benchmark

### PR DESCRIPTION
**Pull request by BoyuanFeng**
This PR adds GenAI layer benchmark. It compares pytorch eager, pytorch compiler, liger, and quack.

It covers all kernels supported by [quack](https://github.com/Dao-AILab/quack?tab=readme-ov-file#kernels-) (CrossEntropy Fwd/Bwd, Softmax Fwd/Bwd, RMSNorm Fwd/Bwd, LayerNorm Fwd) and LayerNormBwd.

## Motivations

- Many OSS users asked how to properly benchmark torch.compile generated kernels. One common error is to compile a kernel/layer for one shape (e.g., batch size=1) and benchmark for another shape (e.g., batch size = 1024), which leads to bad performance. This provides an simple & clear example for proper benchmark.
- We recently added GenAI model benchmark (based on [vLLM](https://hud.pytorch.org/benchmark/llms?repoName=vllm-project%2Fvllm)). But it's usually hard to optimize models directly due to complexity. Layer benchmarks are easier to reason and optimize.

## Key Settings

- Avoid reusing a kernel specializing on 1 shape for benchmark on another shape.
```python
torch._dynamo.config.automatic_dynamic_shapes = False
# Needed since changing args to function causes recompiles
torch._dynamo.config.recompile_limit = 1000000
```

- For forward, people may mark batch size as dynamic to avoid runtime recompilation. We respect the setting in this kernel-level benchmark.
```
torch._dynamo.mark_dynamic(x, 0)
```

GPU: H100 (devvm006.dkl0)

Results: [P1874246170](https://www.internalfb.com/phabricator/paste/view/P1874246170)

Note: for numerical accuracy, we use the default tolerance of torch.testing.assert_close (i.e., for `torch.bfloat16`, use rtol `1.6e-2` and atol `1e-5`). It shows numerical issues for some backends and kernels.

Next step is to add roofline analysis, add to ci for checking regression, cover more GenAI Kernels, and include GenAI Layers for common fusion patterns.

CrossEntropyBackward_bench
CrossEntropyForward_bench
LayerNormBackward_bench
LayerNormForward_bench
RMSNormBackward_bench
RMSNormForward_bench
SoftmaxBackward_bench
SoftmaxForward_bench

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Might also be interesting to add in addition to liger:
- https://github.com/Dao-AILab/quack
- https://github.com/Dao-AILab/flash-attention/tree/main/flash_attn/ops/triton

?

**Comment by BoyuanFeng** [comment]
@vadimkantorov yes. quack is added. thanks for letting me know flash_attn also has ops besides attention

**Comment by vadimkantorov** [comment]
There is also some triton-bench.com / triton-bench.ai effort: https://youtu.be/5e1YKqsP8i8?t=2123

https://github.com/pytorch-labs/tritonbench/

Also, would be maybe good to add here the vanilla examples from triton

File: benchmarks/dynamo/genai_kernels/benchmark.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    SoftmaxBackward,
+    SoftmaxForward,
+)
+
+import torch
+
+
+torch._dynamo.config.automatic_dynamic_shapes = False
+# Needed since changing args to function causes recompiles
+torch._dynamo.config.recompile_limit = 1000000
```
Comment (yf225) [inline benchmarks/dynamo/genai_kernels/benchmark.py]: maybe we can call `torch._dynamo.reset()` after each benchmark? not sure if it has the same effect

**Review by yf225 — APPROVED** [review]

File: benchmarks/dynamo/genai_kernels/benchmark.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+torch._inductor.config.force_disable_caches = True
```
Comment (BoyuanFeng) [inline benchmarks/dynamo/genai_kernels/benchmark.py]: force recompilation for each size

**Review by eellison — APPROVED** [review]
How does this relate to tritonbench? cc @findhao

File: benchmarks/dynamo/genai_kernels/benchmark.py (hunk 1 above)
Comment (BoyuanFeng) [reply benchmarks/dynamo/genai_kernels/benchmark.py]: seems we only need `torch._dynamo.config.automatic_dynamic_shapes = False` and large recompile_limit. Maybe better to be explicit than `torch._dynamo.reset()`?

**Comment by BoyuanFeng** [comment]
This benchmark a) covers genai kernels and will be extended to layers/some common fusion patterns. b) compares inductor against other DSL (e.g., quack, liger, helion, and more). So we can track regression and motivate optimizations.

Triton Bench covers hand-tuned kernels but does not cover layers. It also focuses on triton and may not compare against other DSL.

**Comment by FindHao** [comment]
> How does this relate to tritonbench? cc @FindHao

I guess there are some duplications. @xuzhao9 has added two ops to tritonbench https://github.com/pytorch-labs/tritonbench/pull/289 . Considering TritonBench has been integrated to many internal testings, nightly runs, different version checks etc., I feel it's better to add ops directly to tritonbench.

**Comment by FindHao** [comment]
> a) covers genai kernels and will be extended to layers/some common fusion patterns.

tritonbench covers many kernels especially some internal kernels in internal version. I agree tritonbench doesn't support layers or fusion patterns right now. 

>b) compares inductor against other DSL (e.g., quack, liger, helion, and more)

tritonbench supports comparisons with eager, inductor, liger, quack(2 for now), helion(I remember Xu mentioned somewhere), jax, ThunderKittens, amd iter etc. see dependencies in this file https://github.com/pytorch-labs/tritonbench/blob/main/install.py

**Comment by eellison** [comment]
I know we've talked about showing tritonbench in dashboard (cc @drisspg) - until that's there I think the reality is folks may create other benchmarks.

**Comment by FindHao** [comment]
@xuzhao9 I remember you mentioned there are already some internal benchmark dashboard, right? please correct me if I was wrong. 
@eellison I see. but if it is only for visualization, then we can just have a visualization part. I'm thinking, in long term, those divergences may lead to many more following up duplication works.

**Comment by FindHao** [comment]
I'm not trying to push back this PR, just want to point out the partial duplications and possible effort waste. I'm totally ok to have multiple benchmarks for different needs.

**Comment by BoyuanFeng** [comment]
Offline discussed with @FindHao and decided "GenAI Layer Benchmark" fits better.

**Comment by FindHao** [comment]
yeah, I feel it's more like gptfast to have some reference implementations rather than a benchmark framework.

**Comment by vadimkantorov** [comment]
Some more kernels in https://github.com/pytorch-labs/applied-ai

File: benchmarks/dynamo/genai_layers/kernels.py [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+
+    def compiled(self, args, kwargs=None) -> Any:
+        assert kwargs is None
+        x, w = args
+
+        # Mark batch size as dynamic for realistic workload
+        torch._dynamo.mark_dynamic(x, 0)
+
+        compiled_layernorm = torch.compile(
+            self.layernorm_ref, mode="max-autotune-no-cudagraphs"
```
Comment (zou3519) [inline benchmarks/dynamo/genai_layers/kernels.py:529]: It would be nice to have a way to run the benchmarks without max-autotune.

A lot of users don't use max-autotune because it is prohibitively expensive, so I'd be curious what the numbers are without it.

File: benchmarks/dynamo/genai_layers/utils.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+import torch
+from torch._inductor.runtime.benchmarking import benchmarker
+
+
+def benchmark_kernel_in_milliseconds(func: Callable, *args, **kwargs) -> float:
+    # warmup
+    for _ in range(5):
+        func(*args, **kwargs)
+    return benchmarker.benchmark_gpu(lambda: func(*args, **kwargs))
```
Comment (zou3519) [inline benchmarks/dynamo/genai_layers/utils.py:16]: nit: we should probably enforce no recompiles after warm-up? Just as a sanity check... (it's bad if these do have recompiles)
```
for _ in range(5):
    func(*args, **kwargs)
with torch.compiler.set_stance("fail_on_recompile"):
    return benchmarker.benchmark_gpu(lambda: func(*args, **kwargs))
```

File: benchmarks/dynamo/genai_layers/utils.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        for backend in self.available_backends:
+            if backend == "eager":
+                continue
+            try:
+                torch.testing.assert_close(res[backend], gold)
+                for t, gold_t in zip(res[backend], gold):
+                    if t.requires_grad:
+                        torch.testing.assert_close(t.grad, gold_t.grad)
+                print(
+                    f"Accuracy check \033[92m✓ succeed\033[0m for {backend} backend on {self.name} kernel"
```
Comment (zou3519) [inline benchmarks/dynamo/genai_layers/utils.py:115]: nit: what are the `\033[92m✓`? Is that a color?

Depending if someone is running this in a terminal with or without colors, it's better to not print the colors if the terminal doesn't support it

File: benchmarks/dynamo/genai_layers/kernels.py [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+
+        # Mark batch size as dynamic for realistic workload
+        torch._dynamo.mark_dynamic(x, 0)
+        torch._dynamo.mark_dynamic(target, 0)
+
+        # Need `lambda` otherwise torch.compile will not trace the function.
+        # More discussion: https://github.com/pytorch/pytorch/issues/158455
+        compiled_cross_entropy = torch.compile(
+            lambda x, target: F.cross_entropy(x, target, reduction="none"),
+            mode="max-autotune-no-cudagraphs",
```
Comment (zou3519) [inline benchmarks/dynamo/genai_layers/kernels.py:55]: nit: fullgraph=True, unless these aren't fullgraph-able
